ctpolicy/ctpolicy.go

```diff
@@ -0,0 +1,124 @@
+package ctpolicy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	pubpb "github.com/letsencrypt/boulder/publisher/proto"
+)
+
+// CTPolicy is used to hold information about SCTs required from various
+// groupings
+type CTPolicy struct {
+	pub    core.Publisher
+	groups [][]cmd.LogDescription
+	log    blog.Logger
+}
+
+// New creates a new CTPolicy struct
+func New(pub core.Publisher, groups [][]cmd.LogDescription, log blog.Logger) *CTPolicy {
+	return &CTPolicy{
+		pub:    pub,
+		groups: groups,
+		log:    log,
+	}
+}
+
+type result struct {
+	sct core.SCTDER
+	err error
+}
+
+// race submits an SCT to each log in a group and waits for the first response back,
+// once it has the first SCT it cancels all of the other submissions and returns.
+// It allows up to len(group)-1 of the submissions to fail as we only care about
+// getting a single SCT.
+func (ctp *CTPolicy) race(ctx context.Context, cert core.CertDER, group []cmd.LogDescription) (core.SCTDER, error) {
+	results := make(chan result, len(group))
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for _, l := range group {
+		go func(l cmd.LogDescription) {
+			sct, err := ctp.pub.SubmitToSingleCTWithResult(subCtx, &pubpb.Request{
+				LogURL:       &l.URI,
+				LogPublicKey: &l.Key,
+				Der:          cert,
+			})
+			if err != nil {
+				// Only log the error if it is not a result of canceling subCtx
+				if err != context.Canceled {
+					ctp.log.Warning(fmt.Sprintf("ct submission to %q failed: %s", l.URI, err))
+				}
+				results <- result{err: err}
+				return
+			}
+			results <- result{sct: sct.Sct}
+		}(l)
+	}
+
+	numErr := 0
+	for {
```

Comment:
Might as well make this a more traditional `for` loop with a loop variable, since that's what `numErr` acts as. Then instead of `if numErr == len(group)` we can just fall out the bottom of the loop and return the "all submissions for group failed" error there.